dipy/reconst/ivim.py

```diff
@@ -0,0 +1,157 @@
+        self.args = args
+        self.kwargs = kwargs
+        self.min_signal = self.kwargs.pop('min_signal', None)
+        if self.min_signal is not None and self.min_signal <= 0:
+            e_s = "The `min_signal` key-word argument needs to be strictly"
+            e_s += " positive."
+            raise ValueError(e_s)
+
+    def fit(self, data, mask=None):
+        """ Fit method of the Ivim model class
+
+        Parameters
+        ----------
+        data : array
+            The measured signal from one voxel.
+
+        mask : array
+            A boolean array used to mark the coordinates in the data that
+            should be analyzed that has the shape data.shape[:-1]
+
+        """
+        if mask is None:
+            # Flatten it to 2D either way:
+            data_in_mask = np.reshape(data, (-1, data.shape[-1]))
+        else:
+            # Check for valid shape of the mask
+            if mask.shape != data.shape[:-1]:
+                raise ValueError("Mask is not the same shape as data.")
+            mask = np.array(mask, dtype=bool, copy=False)
+            data_in_mask = np.reshape(data[mask], (-1, data.shape[-1]))
+
+        if self.min_signal is None:
+            min_signal = _min_positive_signal(data)
+        else:
+            min_signal = self.min_signal
+
+        data_in_mask = np.maximum(data_in_mask, min_signal)
+        # Lookup design matrix for vectorization ?
+        params_in_mask = self.fit_method(data_in_mask, self.gtab)
+
+        if mask is None:
+            out_shape = data.shape[:-1] + (-1, )
+            ivim_params = params_in_mask.reshape(out_shape)
+        else:
+            ivim_params = np.zeros(data.shape[:-1] + (4,))
+            ivim_params[mask, :] = params_in_mask
+
+        return IvimFit(self, ivim_params)
+
+
+class IvimFit(object):
+
+    def __init__(self, model, model_params):
+        """ Initialize a IvimFit class instance.
+            Parameters
+            ----------
+            The model parameters are S0, f, D_star, D
+        """
+        self.model = model
+        self.model_params = model_params
+
+    def __getitem__(self, index):
+        model_params = self.model_params
+        N = model_params.ndim
+        if type(index) is not tuple:
+            index = (index,)
+        elif len(index) >= model_params.ndim:
+            raise IndexError("IndexError: invalid index")
+        index = index + (slice(None),) * (N - len(index))
+        return type(self)(self.model, model_params[index])
+
+    @property
+    def shape(self):
+        return self.model_params.shape[:-1]
+
+
+def nlls_fit(data, gtab, jac=False):
+    """
+    Fit the ivim params using non-linear least-squares.
+
+    Parameters
+    ----------
+    data : array ([X, Y, Z, ...], g)
+        Data or response variables holding the data. Note that the last
+        dimension should contain the data. It makes no copies of data.
+
+    jac : bool
```

Comment:
If we replace `curve_fit` with another function (e.g., `scipy.optimize.leastsq`), it might be worth figuring out what the Jacobian function is. It might speed up the calculations quite a bit.